File: results/2026-04-17_004319/tasks/15-test-results-aggregator/typescript-bun-opus47-200k-medium/generated-code/aggregator.ts
// Test-result aggregator: parses JUnit XML + JSON, aggregates across runs,
// finds flaky tests, and renders a GitHub Actions markdown summary.
//
// Design notes:
// - No XML/JSON-schema library dependencies; we use small regex-based parsers
//   scoped to the subsets we care about so the script runs on a vanilla Bun install.
// - All parsing functions accept a `source` label so errors can point back to
//   the offending file path even though parsers themselves are string-based.

export type Status = "passed" | "failed" | "skipped";

export interface TestCase {
  name: string;
  status: Status;
  duration: number; // seconds
}

export interface TestRun {
  source: string;
  duration: number; // seconds
  tests: TestCase[];
}

export interface Totals {
  passed: number;
  failed: number;
  skipped: number;
  total: number;
  duration: number;
}

export interface FlakyEntry {
  name: string;
  passCount: number;
  failCount: number;
}

export interface Aggregate {
  totals: Totals;
  runs: TestRun[];
  flaky: FlakyEntry[];
}

// --- JUnit XML parsing -----------------------------------------------------

// Minimal, tolerant JUnit parser. Pulls <testcase ...>...</testcase> blocks
// plus self-closing variants, and inspects each block for <failure>/<error>/<skipped>.
export function parseJUnit(xml: string, source: string): TestRun {
  if (!/<testsuite[\s>]|<testsuites[\s>]/i.test(xml)) {
    throw new Error(`${source}: does not look like JUnit XML`);
  }

  const tests: TestCase[] = [];
  // Match testcase blocks (self-closing or with body).
  const re = /<testcase\b([^>]*?)(?:\/>|>([\s\S]*?)<\/testcase\s*>)/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(xml)) !== null) {
    const attrs = m[1] ?? "";
    const body = m[2] ?? "";
    const name = attr(attrs, "name") ?? "unknown";
    const classname = attr(attrs, "classname");
    const time = Number(attr(attrs, "time") ?? "0") || 0;
    let status: Status = "passed";
    if (/<failure\b|<error\b/i.test(body)) status = "failed";
    else if (/<skipped\b/i.test(body)) status = "skipped";
    tests.push({
      name: classname ? `${classname}.${name}` : name,
      status,
      duration: time,
    });
  }

  // Sum <testsuite time="..."> attrs for the run total; fall back to case sum.
  let duration = 0;
  const suiteRe = /<testsuite\b([^>]*)>/g;
  let s: RegExpExecArray | null;
  while ((s = suiteRe.exec(xml)) !== null) {
    const t = Number(attr(s[1] ?? "", "time") ?? "");
    if (!Number.isNaN(t)) duration += t;
  }
  if (!duration) duration = tests.reduce((a, t) => a + t.duration, 0);

  return { source, duration, tests };
}

function attr(attrs: string, key: string): string | undefined {
  const m = attrs.match(new RegExp(`\\b${key}\\s*=\\s*"([^"]*)"`, "i"));
  return m ? m[1] : undefined;
}

// --- JSON parsing ----------------------------------------------------------

// Expected shape: { suite?, duration?, tests: [{ name, status, duration }] }
// `status` may be "passed"/"pass"/"ok" etc.; normalize to our Status union.
export function parseJSON(text: string, source: string): TestRun {
  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch (err) {
    throw new Error(`${source}: invalid JSON: ${(err as Error).message}`);
  }
  const obj = data as {
    duration?: number;
    tests?: Array<{ name?: string; status?: string; duration?: number }>;
  };
  if (!obj || !Array.isArray(obj.tests)) {
    throw new Error(`${source}: expected { tests: [...] }`);
  }
  const tests: TestCase[] = obj.tests.map((t, i) => ({
    name: t.name ?? `test-${i}`,
    status: normalizeStatus(t.status),
    duration: Number(t.duration) || 0,
  }));
  const duration =
    Number(obj.duration) || tests.reduce((a, t) => a + t.duration, 0);
  return { source, duration, tests };
}

function normalizeStatus(s: string | undefined): Status {
  const v = (s ?? "").toLowerCase();
  if (v === "pass" || v === "passed" || v === "ok" || v === "success") return "passed";
  if (v === "skip" || v === "skipped" || v === "pending") return "skipped";
  return "failed";
}

// --- File dispatch ---------------------------------------------------------

export async function parseResultFile(path: string): Promise<TestRun> {
  const lower = path.toLowerCase();
  if (!lower.endsWith(".xml") && !lower.endsWith(".json")) {
    throw new Error(`unsupported extension: ${path}`);
  }
  const text = await Bun.file(path).text();
  if (lower.endsWith(".xml")) return parseJUnit(text, path);
  return parseJSON(text, path);
}

// --- Aggregation -----------------------------------------------------------

export function aggregate(runs: TestRun[]): Aggregate {
  const totals: Totals = { passed: 0, failed: 0, skipped: 0, total: 0, duration: 0 };
  for (const run of runs) {
    totals.duration += run.duration;
    for (const t of run.tests) {
      totals[t.status] += 1;
      totals.total += 1;
    }
  }
  return { totals, runs, flaky: findFlaky(runs) };
}

export function findFlaky(runs: TestRun[]): FlakyEntry[] {
  const counts = new Map<string, { pass: number; fail: number }>();
  for (const run of runs) {
    for (const t of run.tests) {
      if (t.status === "skipped") continue;
      const prev = counts.get(t.name) ?? { pass: 0, fail: 0 };
      if (t.status === "passed") prev.pass++;
      else prev.fail++;
      counts.set(t.name, prev);
    }
  }
  const out: FlakyEntry[] = [];
  for (const [name, c] of counts) {
    if (c.pass > 0 && c.fail > 0) {
      out.push({ name, passCount: c.pass, failCount: c.fail });
    }
  }
  out.sort((a, b) => a.name.localeCompare(b.name));
  return out;
}

// --- Markdown rendering ----------------------------------------------------

export function renderMarkdown(agg: Aggregate): string {
  const { totals, runs, flaky } = agg;
  const status = totals.failed > 0 ? "FAILED" : "PASSED";
  const lines: string[] = [];
  lines.push(`# Test Results: ${status}`);
  lines.push("");
  lines.push("## Totals");
  lines.push("");
  lines.push("| Metric | Count |");
  lines.push("| --- | --- |");
  lines.push(`| Passed | ${totals.passed} |`);
  lines.push(`| Failed | ${totals.failed} |`);
  lines.push(`| Skipped | ${totals.skipped} |`);
  lines.push(`| Total | ${totals.total} |`);
  lines.push(`| Duration | ${totals.duration.toFixed(2)}s |`);
  lines.push("");
  lines.push("## Runs");
  lines.push("");
  lines.push("| Source | Passed | Failed | Skipped | Duration |");
  lines.push("| --- | --- | --- | --- | --- |");
  for (const run of runs) {
    const p = run.tests.filter((t) => t.status === "passed").length;
    const f = run.tests.filter((t) => t.status === "failed").length;
    const s = run.tests.filter((t) => t.status === "skipped").length;
    lines.push(
      `| ${run.source} | ${p} | ${f} | ${s} | ${run.duration.toFixed(2)}s |`,
    );
  }
  lines.push("");
  lines.push("## Flaky Tests");
  lines.push("");
  if (flaky.length === 0) {
    lines.push("_None detected._");
  } else {
    lines.push("| Test | Passes | Failures |");
    lines.push("| --- | --- | --- |");
    for (const fl of flaky) {
      lines.push(`| ${fl.name} | ${fl.passCount} | ${fl.failCount} |`);
    }
  }
  lines.push("");
  return lines.join("\n");
}

File: results/2026-05-06_173435/tasks/15-test-results-aggregator/default-opus47-1m-xhigh/generated-code/aggregator.py
"""Test results aggregator.

Parses JUnit XML and JSON test result files, aggregates results across
multiple runs (matrix builds), identifies flaky tests (passed in some runs,
failed in others), and emits a markdown summary suitable for a GitHub Actions
job summary ($GITHUB_STEP_SUMMARY).

Approach:
    parse_junit_xml / parse_json -> List[TestCase]
    Aggregator collects per-run case lists, then computes totals,
    flaky tests, and consistent-failures.
    render_markdown turns the Aggregator into a markdown report.
    main() is the CLI entry point used by the GitHub Actions workflow.

Design notes:
    - Status is normalized to one of {"passed", "failed", "skipped"} so
      flaky-detection is straightforward.
    - A flaky test is one whose normalized status varies between
      "passed" and "failed" across the runs (skipped runs are ignored,
      since "skipped" is intentional, not a flake signal).
    - Errors at parse time raise ValueError with a friendly message
      that names the offending file. This makes CI logs actionable.
"""
from __future__ import annotations

import argparse
import json
import os
import sys
from collections import defaultdict
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable
from xml.etree import ElementTree as ET


# ---------- Domain ----------

@dataclass
class TestCase:
    """A single test case from one run.

    `full_name` is "{suite}.{name}" — a stable identifier we use to
    group across runs for flaky detection.
    """
    suite: str
    name: str
    status: str  # "passed" | "failed" | "skipped"
    duration: float
    failure_message: str | None = None

    @property
    def full_name(self) -> str:
        return f"{self.suite}.{self.name}" if self.suite else self.name


@dataclass
class FlakyResult:
    full_name: str
    pass_count: int
    fail_count: int
    skip_count: int
    sample_failure: str | None


# ---------- Status normalization ----------

# Map common status aliases to our three canonical values.
_STATUS_ALIASES = {
    "passed": "passed", "pass": "passed", "ok": "passed", "success": "passed",
    "successful": "passed",
    "failed": "failed", "fail": "failed", "failure": "failed", "error": "failed",
    "errored": "failed", "broken": "failed",
    "skipped": "skipped", "skip": "skipped", "pending": "skipped",
    "ignored": "skipped",
}


def _normalize_status(raw: str) -> str:
    s = (raw or "").strip().lower()
    if s not in _STATUS_ALIASES:
        # Unknown → treat as failure so we don't silently pass bad data.
        return "failed"
    return _STATUS_ALIASES[s]


# ---------- Parsers ----------

def parse_junit_xml(path: Path) -> list[TestCase]:
    """Parse a JUnit XML file.

    Handles both <testsuites>-rooted and bare <testsuite> documents
    (both are valid in the wild). Each <testcase> becomes one TestCase;
    presence of <failure> or <error> means failed; <skipped> means
    skipped; otherwise passed.
    """
    path = Path(path)
    try:
        tree = ET.parse(path)
    except ET.ParseError as exc:
        raise ValueError(f"Invalid JUnit XML in {path}: {exc}") from exc

    root = tree.getroot()
    suites = (
        list(root.iter("testsuite"))
        if root.tag != "testsuite"
        else [root]
    )

    cases: list[TestCase] = []
    for suite in suites:
        suite_name = suite.attrib.get("name", "")
        for tc in suite.findall("testcase"):
            classname = tc.attrib.get("classname") or suite_name
            name = tc.attrib.get("name", "")
            try:
                duration = float(tc.attrib.get("time", "0") or 0)
            except ValueError:
                duration = 0.0

            failure_el = tc.find("failure")
            error_el = tc.find("error")
            skipped_el = tc.find("skipped")

            if failure_el is not None or error_el is not None:
                status = "failed"
                el = failure_el if failure_el is not None else error_el
                msg = el.attrib.get("message") or (el.text or "").strip() or None
            elif skipped_el is not None:
                status = "skipped"
                msg = None
            else:
                status = "passed"
                msg = None

            cases.append(TestCase(
                suite=classname, name=name, status=status,
                duration=duration, failure_message=msg,
            ))
    return cases


def parse_json(path: Path) -> list[TestCase]:
    """Parse a JSON test results file.

    Expected schema (intentionally simple — a common interchange shape):
        {"results": [
            {"suite": "...", "name": "...", "status": "passed|failed|skipped",
             "duration": 0.42, "message": "optional failure message"},
            ...
        ]}
    """
    path = Path(path)
    try:
        data = json.loads(path.read_text())
    except json.JSONDecodeError as exc:
        raise ValueError(f"Invalid JSON in {path}: {exc}") from exc

    results = data.get("results")
    if not isinstance(results, list):
        raise ValueError(
            f"Invalid JSON in {path}: expected top-level object with 'results' list"
        )

    cases: list[TestCase] = []
    for i, item in enumerate(results):
        if not isinstance(item, dict):
            raise ValueError(f"Invalid JSON in {path}: results[{i}] is not an object")
        try:
            duration = float(item.get("duration", 0) or 0)
        except (TypeError, ValueError):
            duration = 0.0
        cases.append(TestCase(
            suite=str(item.get("suite", "")),
            name=str(item.get("name", "")),
            status=_normalize_status(item.get("status", "")),
            duration=duration,
            failure_message=item.get("message") or item.get("failure_message"),
        ))
    return cases


def parse_file(path: Path) -> list[TestCase]:
    """Dispatch to the right parser by file extension."""
    path = Path(path)
    ext = path.suffix.lower()
    if ext == ".xml":
        return parse_junit_xml(path)
    if ext == ".json":
        return parse_json(path)
    raise ValueError(f"Unsupported file type: {path} (expected .xml or .json)")


# ---------- Aggregation ----------

class Aggregator:
    """Collects per-run test cases and computes summary statistics."""

    def __init__(self) -> None:
        # run_id -> list[TestCase]
        self.runs: dict[str, list[TestCase]] = {}

    def add_run(self, run_id: str, cases: Iterable[TestCase]) -> None:
        self.runs[run_id] = list(cases)

    def all_cases(self) -> Iterable[TestCase]:
        for cases in self.runs.values():
            yield from cases

    def totals(self) -> dict:
        passed = failed = skipped = 0
        duration = 0.0
        for c in self.all_cases():
            duration += c.duration
            if c.status == "passed":
                passed += 1
            elif c.status == "failed":
                failed += 1
            elif c.status == "skipped":
                skipped += 1
        total = passed + failed + skipped
        return {
            "passed": passed, "failed": failed, "skipped": skipped,
            "total": total, "duration": duration, "runs": len(self.runs),
        }

    def _status_by_test(self) -> dict[str, list[TestCase]]:
        """Group cases across runs by full_name."""
        grouped: dict[str, list[TestCase]] = defaultdict(list)
        for c in self.all_cases():
            grouped[c.full_name].append(c)
        return grouped

    def flaky_tests(self) -> list[FlakyResult]:
        """A test is flaky if it has BOTH a passed and a failed run.

        Skipped runs are ignored — being skipped is intentional (e.g. a
        platform-specific exclusion), not a flakiness signal.
        """
        flaky: list[FlakyResult] = []
        for full_name, cases in self._status_by_test().items():
            statuses = [c.status for c in cases]
            pass_count = statuses.count("passed")
            fail_count = statuses.count("failed")
            skip_count = statuses.count("skipped")
            if pass_count > 0 and fail_count > 0:
                sample = next(
                    (c.failure_message for c in cases
                     if c.status == "failed" and c.failure_message),
                    None,
                )
                flaky.append(FlakyResult(
                    full_name=full_name,
                    pass_count=pass_count, fail_count=fail_count,
                    skip_count=skip_count, sample_failure=sample,
                ))
        flaky.sort(key=lambda f: f.full_name)
        return flaky

    def consistent_failures(self) -> list[TestCase]:
        """Tests that failed in every non-skipped run."""
        out: list[TestCase] = []
        for full_name, cases in self._status_by_test().items():
            non_skipped = [c for c in cases if c.status != "skipped"]
            if non_skipped and all(c.status == "failed" for c in non_skipped):
                out.append(non_skipped[0])
        out.sort(key=lambda c: c.full_name)
        return out


# ---------- Markdown rendering ----------

def render_markdown(agg: Aggregator) -> str:
    """Produce a GitHub-Actions-job-summary friendly markdown document."""
    t = agg.totals()
    flaky = agg.flaky_tests()
    failures = agg.consistent_failures()

    overall = "passed" if t["failed"] == 0 and not flaky else "needs attention"
    lines: list[str] = []
    lines.append("# Test Results Summary")
    lines.append("")
    lines.append(f"**Status:** {overall}")
    lines.append("")
    lines.append("## Totals")
    lines.append("")
    lines.append("| Metric | Value |")
    lines.append("|---|---|")
    lines.append(f"| Runs aggregated | {t['runs']} |")
    lines.append(f"| Total tests | {t['total']} |")
    lines.append(f"| Passed | {t['passed']} |")
    lines.append(f"| Failed | {t['failed']} |")
    lines.append(f"| Skipped | {t['skipped']} |")
    lines.append(f"| Total duration | {t['duration']:.2f}s |")
    lines.append("")

    if t["failed"] == 0 and not flaky:
        lines.append("All tests passed across all runs.")
        lines.append("")
        return "\n".join(lines)

    if failures:
        lines.append("## Failures")
        lines.append("")
        lines.append("| Test | Message |")
        lines.append("|---|---|")
        for c in failures:
            msg = (c.failure_message or "").replace("|", "\\|").replace("\n", " ")
            if len(msg) > 120:
                msg = msg[:117] + "..."
            lines.append(f"| `{c.full_name}` | {msg} |")
        lines.append("")

    if flaky:
        lines.append("## Flaky tests")
        lines.append("")
        lines.append("Tests below passed in some runs and failed in others.")
        lines.append("")
        lines.append("| Test | Passed | Failed | Sample failure |")
        lines.append("|---|---|---|---|")
        for f in flaky:
            sample = (f.sample_failure or "").replace("|", "\\|").replace("\n", " ")
            if len(sample) > 80:
                sample = sample[:77] + "..."
            lines.append(
                f"| `{f.full_name}` | {f.pass_count} | {f.fail_count} | {sample} |"
            )
        lines.append("")

    return "\n".join(lines)


# ---------- CLI ----------

def _build_aggregator(paths: list[Path]) -> Aggregator:
    agg = Aggregator()
    for p in paths:
        cases = parse_file(p)
        # Use the file path as the run identifier so each input file is
        # treated as a separate matrix run.
        agg.add_run(str(p), cases)
    return agg


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Aggregate JUnit XML / JSON test results and emit a "
                    "markdown summary."
    )
    parser.add_argument("inputs", nargs="*",
                        help="Paths to test result files (.xml or .json).")
    parser.add_argument("--output", "-o", default="-",
                        help="Output markdown path. '-' = stdout. "
                             "Defaults to $GITHUB_STEP_SUMMARY when set.")
    parser.add_argument("--fail-on-failures", action="store_true",
                        help="Exit 1 when any test failed across all runs "
                             "(after producing the summary). Default: always "
                             "exit 0 on successful summary generation.")
    args = parser.parse_args(argv)

    if not args.inputs:
        print("error: no input files provided", file=sys.stderr)
        return 2

    paths = [Path(p) for p in args.inputs]
    missing = [p for p in paths if not p.exists()]
    if missing:
        print(f"error: missing input file(s): {', '.join(map(str, missing))}",
              file=sys.stderr)
        return 2

    try:
        agg = _build_aggregator(paths)
    except ValueError as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1

    md = render_markdown(agg)

    # If --output not given but $GITHUB_STEP_SUMMARY is set, write there
    # so this drops cleanly into a GH Actions job summary.
    output = args.output
    if output == "-" and os.environ.get("GITHUB_STEP_SUMMARY"):
        output = os.environ["GITHUB_STEP_SUMMARY"]

    if output == "-":
        sys.stdout.write(md)
    else:
        Path(output).write_text(md)

    # Also print key totals to stdout — useful for CI log scraping.
    t = agg.totals()
    print(
        f"Aggregated {t['runs']} run(s): "
        f"{t['passed']} passed, {t['failed']} failed, "
        f"{t['skipped']} skipped, {t['total']} total, "
        f"{t['duration']:.2f}s",
    )
    flaky = agg.flaky_tests()
    if flaky:
        print(f"Flaky tests detected: {len(flaky)}")
        for f in flaky:
            print(f"  - {f.full_name} (passed {f.pass_count}, failed {f.fail_count})")

    # By default the aggregator exits 0 — generating a summary IS the
    # success criterion, and CI surfaces test failures via the markdown
    # itself. Pass --fail-on-failures to bubble test failures up as a
    # non-zero exit (useful when you want the build to break).
    if args.fail_on_failures and t["failed"] > 0 and agg.consistent_failures():
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
